tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -291,60 +512,124 @@ decompress_chunk_begin(CustomScanState *node, EState *estate, int eflags)
 		}
 	}
 
-	initialize_column_state(state);
-
 	node->custom_ps = lappend(node->custom_ps, ExecInitNode(compressed_scan, estate, eflags));
-
-	state->per_batch_context = AllocSetContextCreate(CurrentMemoryContext,
-													 "DecompressChunk per_batch",
-													 ALLOCSET_DEFAULT_SIZES);
 }
 
 static void
-initialize_batch(DecompressChunkState *state, TupleTableSlot *compressed_slot,
-				 TupleTableSlot *decompressed_slot)
+initialize_batch(DecompressChunkState *chunk_state, DecompressBatchState *batch_state,
+				 TupleTableSlot *subslot)
 {
 	Datum value;
 	bool isnull;
 	int i;
-	MemoryContext old_context = MemoryContextSwitchTo(state->per_batch_context);
-	MemoryContextReset(state->per_batch_context);
 
-	state->total_batch_rows = 0;
-	state->current_batch_row = 0;
+	Assert(batch_state->initialized == false);
+
+	/* Batch states can be re-used skip tuple slot creation in that case */
+	if (batch_state->compressed_slot == NULL)
+	{
+		TupleDesc tdesc_sub = CreateTupleDescCopy(subslot->tts_tupleDescriptor);
+		batch_state->compressed_slot = MakeSingleTupleTableSlot(tdesc_sub, subslot->tts_ops);
+	}
```

Comment:
I think I wrote this already somewhere, but can we use a raw pointer to the tupledesc? We don't really have to use reference-counted copies here. It's read-only, and owned by the DecompressChunk node. I remember seeing these tuple operations in profiles, they are slow because they go to the ResourceOwner.

Reply:
I tested this change locally and it works. So, I updated it in the current version of the PR and avoid the calls of `CreateTupleDescCopy`.